include/sparrow/buffer.hpp

```diff
@@ -18,454 +18,870 @@
+        });
     }
 
     template <class T>
-    buffer<T>& buffer<T>::operator=(const buffer<T>& rhs)
+    constexpr void buffer<T>::shrink_to_fit()
     {
-        if (this != &rhs)
+        if (capacity() != size())
         {
-            buffer<T> tmp(rhs);
-            base_type::swap(tmp);
+            buffer(std::make_move_iterator(begin()), std::make_move_iterator(end()), get_allocator()).swap(*this);
         }
-        return *this;
+    }
+    
+    template <class T>
+    constexpr void buffer<T>::swap(buffer& rhs)
+    {
+        std::swap(this->get_data(), rhs.get_data());
     }
 
     template <class T>
-    buffer<T>::buffer(buffer&& rhs)
-        : base_type{rhs.data(), rhs.size()}
+    template <class F>
+    constexpr void buffer<T>::resize_impl(size_type new_size, F&& initializer)
     {
-        rhs.reset(nullptr, 0u);
+        if (new_size > size())
+        {
+            const size_t nb_init = new_size - size();
+            if (new_size <= capacity())
+            {
+                initializer(nb_init);
+            }
+            else
+            {
+                reserve(new_size);
+                initializer(nb_init);
+            }
+        }
+        else if (new_size < size())
+        {
+            erase_at_end(get_data().p_begin + new_size);
+        }
+    }
+    
+    template <class T>
+    template <class It>
+    constexpr void buffer<T>::assign_range_impl(It first, It last, std::forward_iterator_tag)
+    {
+        const size_type sz = size();
+        const size_type len = std::distance(first, last);
+        if (len > capacity())
+        {
+            check_init_length(len, get_allocator());
+            pointer p = allocate_and_copy(len, first, last);
+            destroy(get_data().p_begin, get_data().p_end, get_allocator());
+            this->deallocate(get_data().p_begin, capacity());
+            this->assign_storage(p, len, len);
+        }
+        else if (sz >= len)
+        {
+            pointer p = std::copy(first, last, get_data().p_begin);
+            erase_at_end(p);
+        }
+        else
+        {
+            It mid = first;
+            std::advance(mid, sz);
+            std::copy(first, mid, get_data().p_begin);
+            get_data().p_end = copy_initialize(mid, last, get_data().p_end, get_allocator());
+        }
     }
 
     template <class T>
-    buffer<T>& buffer<T>::operator=(buffer<T>&& rhs)
+    constexpr void buffer<T>::erase_at_end(pointer p)
     {
-        base_type::swap(rhs);
-        return *this;
+        destroy(p, get_data().p_end, get_allocator());
+        get_data().p_end = p;
     }
 
     template <class T>
-    void buffer<T>::resize(size_type n)
+    template <class It>
+    constexpr auto buffer<T>::allocate_and_copy(size_type n, It first, It last) -> pointer
     {
-        // TODO: add capacity, resize if growing only and define a shrink_to_fit method
-        if (n != base_type::size())
+        pointer p = this->allocate(n);
```

Comment:
Must it be present in the exception handling bellow?

Reply:
No, the exception handling is to make sure the memory is deallocated if the initialization throws. If the allcoation throws, we have nothing to deallocate and we let the user decide what to do.